sdk/core/azure-core/src/http/challenge_based_authentication_policy.cpp

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "azure/core/http/policies/policy.hpp"
+#include <chrono>
+#include <vector>
+
+using Azure::Core::Context;
+using namespace Azure::Core::Http;
+using namespace Azure::Core::Http::Policies;
+using namespace Azure::Core::Http::Policies::_internal;
+using namespace Azure::Core::Http::Policies::_detail;
+
+std::unique_ptr<RawResponse> ChallengeBasedAuthenticationPolicy::Send(
+    Request& request,
+    NextHttpPolicy nextPolicy,
+    Context const& context) const
+{
+  request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+  auto result = nextPolicy.Send(request, context);
+  auto const& headers = result->GetHeaders();
+  auto const& rawDataHeader = headers.find("www-authenticate");
+  // Only re-run when Unauthozied AND www-authenticate header in the response
+  if (result->GetStatusCode() == HttpStatusCode::Unauthorized && rawDataHeader != headers.end())
+  {
+    ChallengeParameters challenge(rawDataHeader->second);
+    Credentials::TokenRequestContext tokenRequestContext;
+
+    tokenRequestContext.TenantId = std::move(challenge.TenantId);
+    tokenRequestContext.AuthorizationUri = std::move(challenge.AuthorizationUri);
+    tokenRequestContext.Scopes = std::move(challenge.Scopes);
+
+    {
+      std::lock_guard<std::mutex> lock(m_accessTokenMutex);
+      m_accessToken = m_credential->GetToken(tokenRequestContext, context);
+      request.SetHeader("authorization", "Bearer " + m_accessToken.Token);
+    }
+    return nextPolicy.Send(request, context);
+  }
+
+  // Did not re-send on any other case
+  return result;
+}
+/*
+ * the raw value ( the value of the "www-authenticate" header is in the following format
+ * "Bearer authorization/authorization_uri=[value] resource/scope=[value]";
+ */
+ChallengeParameters::ChallengeParameters(std::string const& headerValue)
+{
+  if (headerValue.rfind(_detail::BearerName) == 0)
+  {
+    Schema = _detail::BearerName;
+
+    std::vector<std::string> const parts = GetParts(headerValue, _detail::SpaceSeparator);
```

Comment:
`GetParts` is basically an `Split()` function.  
It would be useful if we were trying to mutate any of the values parsed out of the source string (headerValue).
However, all we need is to loop `headerValue` from `begin()` to `end()`.  Using 2 iterators to move within the source string will avoid allocating the `vector<string>`.

Reply:
for one thing that is similar to voodoo magic, it is a lot harder to debug, trace, test. I honestly fail to see the benefit of it. as for speed i do not buy it , we use vectors all over the place, without any speed issues.

Reply:
> as for speed i do not buy it 

It's easy to see.  One single loop will give use O(n), while currently you are doing O(n + m * v), where `n` is the source string, then `m` is the first getParts().size() and `v` is the next getParts().size() called for each part.

Also, see how `SetHeader()` is implemented:
```cpp
static void inline SetHeader(
          Azure::Core::Http::RawResponse& response,
          uint8_t const* const first,
          uint8_t const* const last)
      {
        // get name and value from header
        auto start = first;
        auto end = std::find(start, last, ':');

        if (end == last)
        {
          throw std::invalid_argument("Invalid header. No delimiter ':' found.");
        }

        // Always toLower() headers
        auto headerName
            = Azure::Core::_internal::StringExtensions::ToLower(std::string(start, end));
        start = end + 1; // start value
        while (start < last && (*start == ' ' || *start == '\t'))
        {
          ++start;
        }

        end = std::find(start, last, '\r');
        auto headerValue = std::string(start, end); // remove \r

        response.SetHeader(headerName, headerValue);
      }
    };
```

And how `CreateHTTPResponse()` is implemeted:
```cpp
static std::unique_ptr<RawResponse> CreateHTTPResponse(
    uint8_t const* const begin,
    uint8_t const* const last)
{
  // set response code, HTTP version and reason phrase (i.e. HTTP/1.1 200 OK)
  auto start = begin + 5; // HTTP = 4, / = 1, moving to 5th place for version
  auto end = std::find(start, last, '.');
  auto majorVersion = std::stoi(std::string(start, end));

  start = end + 1; // start of minor version
  end = std::find(start, last, ' ');
  auto minorVersion = std::stoi(std::string(start, end));

  start = end + 1; // start of status code
  end = std::find(start, last, ' ');
  auto statusCode = std::stoi(std::string(start, end));

  start = end + 1; // start of reason phrase
  end = std::find(start, last, '\r');
  auto reasonPhrase = std::string(start, end); // remove \r

  // allocate the instance of response to heap with shared ptr
  // So this memory gets delegated outside CurlTransport as a shared_ptr so memory will be
  // eventually released
  return std::make_unique<RawResponse>(
      static_cast<uint16_t>(majorVersion),
      static_cast<uint16_t>(minorVersion),
      HttpStatusCode(statusCode),
      reasonPhrase);
}
```

Can you point to any implementation from the SDK where we have chosen to allocate heap components (vectors/strings) instead of using iterators? so it is easier to debug, trace and test? (which should not be related at all)